explorer/testdata/class/virtual_method.carbon

```diff
@@ -0,0 +1,48 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// RUN: %{explorer-run}
+// RUN: %{explorer-run-trace}
+// CHECK:STDOUT: c.Foo() -> 1
+// CHECK:STDOUT: c.Bar() -> 2
+// CHECK:STDOUT: d.Foo() -> 3
+// CHECK:STDOUT: d.Bar() -> 4
+// CHECK:STDOUT: cc.Foo() -> 3
+// CHECK:STDOUT: cc.Bar() -> 2
+// CHECK:STDOUT: result: 0
+
+package ExplorerTest api;
+
+base class C {
```

Comment:
Can you add as another test:

```
base class A {}

class B extends A {
  virtual fn Foo[self: Self]() -> i32 { return 0; }
}

fn Main() -> i32 {
  var b: B = {};
  var bp: B* = &b;
  var ap: A* = &a;
  // Shouldn't look into the vtable for B.
  return ap->Foo();
}
```

Note, I'm suspicious this may _work_ because it'll look into the vtable of B. Addressing another comment ("regular function lookup needs to find a virtual function before it can conclude that a vtable lookup should occur for that function.") would address this if there is a bug here though, as the lookup for Foo in A's API would fail, so the vtable would never be consulted.

Reply:
Added here: [explorer/testdata/class/fail_virtual_method_absent.carbon](https://github.com/carbon-language/carbon-lang/pull/2462/files#diff-4819c57259bc06a0212b68da74d84697ba665eb8cc33dfe6fce48c973a5db948).

Modified logic accordingly, looking better.

However, it did highlight an issue with `self`. Added a new test for that, and working on it.